File: 106. Construct Binary Tree from Inorder and Postorder Traversal/main.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {
	ans := buildTree([]int{4, 2, 5, 1, 6, 3, 7}, []int{4, 5, 2, 6, 7, 3, 1})
	fmt.Println(ans)
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(inorder []int, postorder []int) *TreeNode {
	n := len(inorder) - 1
	return traverse(inorder, postorder, 0, n, 0, n)
}

func traverse(inorder []int, postorder []int, instart int, inend int, poststart int, postend int) *TreeNode {
	if poststart > postend {
		return nil
	}

	root := postorder[postend]
	rootPos := -1
	for i, num := range inorder {
		if num == root {
			rootPos = i
			break
		}
	}
	leftSize := rootPos - instart

	node := TreeNode{
		Val:   root,
		Left:  traverse(inorder, postorder, instart, rootPos-1, poststart, poststart+leftSize-1),
		Right: traverse(inorder, postorder, rootPos+1, inend, poststart+leftSize, postend-1),
	}

	return &node
}

File: 106. Construct Binary Tree from Inorder and Postorder Traversal/main.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
from typing import List, Optional


class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if not postorder or len(postorder) == 0:
            return None

        root = TreeNode(postorder[-1])
        stack = []
        stack.append(root)
        inorderIdx = len(inorder) - 1

        for i in range(len(postorder) - 2, -1, -1):
            postOrderVal = postorder[i]
            node = stack[-1]
            if node.val != inorder[inorderIdx]:
                node.right = TreeNode(postOrderVal)
                stack.append(node.right)
            else:
                while len(stack) > 0 and stack[-1].val == inorder[inorderIdx]:
                    node = stack.pop()
                    inorderIdx-=1
                node.left = TreeNode(postOrderVal)
                stack.append(node.left)

        return root